src/Language/PureScript/Linter/Exhaustive.hs

```diff
@@ -246,69 +258,83 @@ checkExhaustive hasConstraint env mn numArgs cas = makeResult . first nub $ fold
-  onExpr isP (TypeClassDictionaryConstructorApp _ e) = onExpr isP e
-  onExpr isP (Accessor _ e) = onExpr isP e
-  onExpr isP (ObjectUpdate o es) = onExpr isP o >> mapM_ (onExpr isP . snd) es
-  onExpr isP (Abs _ e) = onExpr isP e
-  onExpr isP (App e1 e2) = onExpr isP e1 >> onExpr isP e2
-  onExpr isP (IfThenElse e1 e2 e3) = onExpr isP e1 >> onExpr isP e2 >> onExpr isP e3
-  onExpr isP (Case es cas) = checkExhaustive isP env mn (length es) cas >> mapM_ (onExpr isP) es >> mapM_ (onCaseAlternative isP) cas
-  onExpr isP (TypedValue _ e ty) = onExpr (isP || hasPartialConstraint ty) e
-  onExpr isP (Let ds e) = mapM_ onDecl ds >> onExpr isP e
-  onExpr isP (PositionedValue pos _ e) = censor (addHint (PositionedError pos)) (onExpr isP e)
-  onExpr _ _ = return ()
-
-  onCaseAlternative :: Bool -> CaseAlternative -> m ()
-  onCaseAlternative isP (CaseAlternative _ (Left es)) = mapM_ (\(e, g) -> onExpr isP e >> onExpr isP g) es
-  onCaseAlternative isP (CaseAlternative _ (Right e)) = onExpr isP e
-
-  hasPartialConstraint :: Type -> Bool
-  hasPartialConstraint (ConstrainedType cs _) = any (go . fst) cs
+      tellRedundant = tell . errorMessage . uncurry OverlappingPattern . second null . splitAt 5 $ bss'
+      tellIncomplete = tell . errorMessage $ IncompleteExhaustivityCheck
+
+  -- | We add a Partial constraint by adding a call to the following identity function:
+  --
+  -- partial :: forall a. Partial => a -> a
+  --
+  -- The binder information is provided so that it can be embedded in the constraint,
+  -- and then included in the error message.
+  addPartialConstraint :: ([[Binder]], Bool) -> Expr -> Expr
+  addPartialConstraint (bss, complete) e =
+      Let [ partial ] (App (Var (Qualified Nothing (Ident C.__unused))) e)
     where
-    go :: Qualified (ProperName 'ClassName) -> Bool
-    go qname
-      | qname == partialClass = True
-      | otherwise =
-          case qname `M.lookup` typeClasses env of
-            Just ([], _, cs') -> any (go . fst) cs'
-            _ -> False
-    partialClass :: Qualified (ProperName 'ClassName)
-    partialClass = primName "Partial"
-  hasPartialConstraint _ = False
+      partial :: Declaration
+      partial = ValueDeclaration (Ident C.__unused)
+                                 Private
+                                 []
+                                 (Right (TypedValue True (Abs (Left (Ident "x"))
+                                                              (Var (Qualified Nothing (Ident "x"))))
+                                                         ty))
+
+      ty :: Type
+      ty = ForAll "a" (ConstrainedType [ Constraint (Qualified (Just (ModuleName [ProperName C.prim]))
+                                                               (ProperName "Partial"))
+                                                    []
+                                                    (Just (PartialConstraintData (map (map prettyPrintBinderAtom) bss) complete))
```

Comment:
It introduces a circular dependency between the `Binders` and `Types` modules. We talked about it a bit, and since this is basically a temporary solution until we have constraint kinds and programmable error messages, using a string didn't seem so bad.